### PR TITLE
Use more Ruby for basic linter

### PR DESCRIPTION
**Pull request by splattael**
## WIP! Needs review!

__Rebased and squashed down. Details are from the commit messages__

Adds a basic linter (`ROM::Lint::Linter`) that wraps up the common needs out of a plain Ruby linter: a list of lints, named like test methods (lint_*), a way to yield to each lint method, and a method to perform a specific lint returning `true` so they are assertion friendly.

Adds specific linters for a repository and a dataset. These can be used by repository adapter authors to verify if they comply with our api.

Updates the Minitest/TestUnit module to use the new `ROM::Lint::Test` modules (`TestRepository` & `TestEnumerableDataset`) to get access to the test method versions of the lint tests.

Use new repository feature linters to create rspec shared examples for use in an rspec suite.

Add shared features to `memory/dataset_spec` and create new `repository_spec` so our linters are running as part of our normal spec suite.

### CODE REVIEW

**Comment by solnic** [comment]
Please see my commits in the reverted [commit](https://github.com/rom-rb/rom/commit/f59011fccc06d5ce552891261463ae31851cc28d)

**Comment by splattael** [comment]
Just an idea. How about define lint methods like that:

```ruby
lint "schemes returns any supported scheme" do
  assert "#{adapter}#schemes must return at least one supported URI scheme" do
    adapter.schemes.any?
  end
end
```

Too fancy?

**Comment by splattael** [comment]
I've merged with master. We can squash afterwards, right?

**Comment by solnic** [comment]
Si. I guess this needs some updates after repo/adapter changes.

**Comment by splattael** [comment]
After the Adapter/Repository collapse: Are we still linting adapters; or now repositories? :dancers:

**Comment by solnic** [comment]
I think now "Adapter" describes a library that provides support for a specific datastore. So, yes, this is still linting for adapters, more specifically lint tests for base adapter objects: `Repository` and `Dataset`

**Comment by splattael** [comment]
Merged with fast-moving master.

I'm still not sure if I'm mistaken in 69e383b by ditching `Adapter` namespace.

**Comment by elskwid** [comment]
@splattael as I mentioned in chat, this is going to require a helluva rebase. Are you okay with us losing some detail in the commits? With the Adapter removal it's going to be very difficult to keep it all straight.

**Comment by splattael** [comment]
@elskwid Sure, no problem. Squash it :)

**Comment by elskwid** [comment]
@splattael and @solnic (@aflatter too) - this is ready for a good look-over. :eyes:

**Comment by phlipper** [comment]
This is a really nice little changset, I'd only recommend adding some context to the README.

File: lib/rom/lint/repository.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return if repository.schemes.instance_of? Array
+
+        complain "#{repository}#schemes must return an array with supported schemes"
+      end
+
+      def lint_schemes_returns_any_supported_scheme
+        return if repository.schemes.any?
+
+        complain "#{repository}#schemes must return at least one supported scheme"
+      end
```
Comment (solnic) [inline lib/rom/lint/repository.rb]: scheme-related checks can go away given that #120 removes it

File: spec/unit/rom/memory/repository_spec.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+require 'spec_helper'
+require 'rom/lint/spec'
+require 'rom/memory.rb'
```
Comment (splattael) [inline spec/unit/rom/memory/repository_spec.rb]: `.rb` extension is not needed
`require 'rom/memory'`

Comment (elskwid) [inline spec/unit/rom/memory/repository_spec.rb]: Good catch!

**Pull request merged**